### PR TITLE
add support for http basic auth

### PR DESCRIPTION
**Pull request by hmoqhim**
reference  #1059

This is to add support for http basic authentication on parser-server

### CODE REVIEW

File: src/middlewares.js [hunk 1]
```diff
@@ -144,6 +146,33 @@ function handleParseHeaders(req, res, next) {
     });
 }
 
+function httpAuth(req) {
+  var appId, masterKey, javascriptKey;
+  var credentialsRegExp = /^ *(?:[Bb][Aa][Ss][Ii][Cc]) +([A-Za-z0-9\-\._~\+\/]+=*) *$/
+  var appIdKeyRegExp = /^([^ :]*):([Jj][Aa][Vv][Aa][Ss][Cc][Rr][Ii][Pp][Tt][-][Kk][Ee][Yy]=)?(.*)/
```
Comment (flovilmart) [inline src/middlewares.js]: could we use case insensitive search to simplify the regexp?

File: src/middlewares.js [hunk 2]
```diff
@@ -144,6 +146,33 @@ function handleParseHeaders(req, res, next) {
...
+ // get header
+  var header = (req.req || req).headers.authorization
```
Comment (flovilmart) [inline src/middlewares.js]: I'd check for that earlier and early exit if `req.headers.authorization` is not defined

File: src/middlewares.js [hunk 3]
```diff
@@ -144,6 +146,33 @@ function handleParseHeaders(req, res, next) {
...
+  // parse header
+  var match = credentialsRegExp.exec(header || '')
```
Comment (flovilmart) [inline src/middlewares.js]: may be just check that `indexOf('Basic: ') == 0` instead of the convoluted regexp

File: src/middlewares.js [hunk 4]
```diff
@@ -144,6 +146,33 @@ function handleParseHeaders(req, res, next) {
...
+  if (match) {
+    // decode user pass
+    var appIdKeyMatch = appIdKeyRegExp.exec(decodeBase64(match[1]))
+	 if (appIdKeyMatch) {
+      appId = appIdKeyMatch[1];
+      if (appIdKeyMatch[2])
```
Comment (flovilmart) [inline src/middlewares.js]: use `{` or put on one line

File: src/middlewares.js [hunk 5]
```diff
@@ -144,6 +146,33 @@ function handleParseHeaders(req, res, next) {
...
+  var credentialsRegExp = /^ *(?:[Bb][Aa][Ss][Ii][Cc]) +([A-Za-z0-9\-\._~\+\/]+=*) *$/
+  var appIdKeyRegExp = /^([^ :]*):([Jj][Aa][Vv][Aa][Ss][Cc][Rr][Ii][Pp][Tt][-][Kk][Ee][Yy]=)?(.*)/
+ // get header
+  var header = (req.req || req).headers.authorization
+  // parse header
+  var match = credentialsRegExp.exec(header || '')
+  if (match) {
+    // decode user pass
+    var appIdKeyMatch = appIdKeyRegExp.exec(decodeBase64(match[1]))
+	 if (appIdKeyMatch) {
```
Comment (flovilmart) [inline src/middlewares.js]: align correctly

File: src/middlewares.js [hunk 6]
```diff
@@ -17,13 +17,15 @@ function handleParseHeaders(req, res, next) {
   var mountPath = req.originalUrl.slice(0, mountPathLength);
   var mount = req.protocol + '://' + req.get('host') + mountPath;
 
+  var basicAuth = httpAuth(req);
+
   var info = {
-    appId: req.get('X-Parse-Application-Id'),
+    appId: basicAuth.appId || req.get('X-Parse-Application-Id'),
```
Comment (flovilmart) [inline src/middlewares.js]: not sure I like mixing those two code paths together. 
Maybe httpAuth should not return anything if nothing is set.

This path may be prone to mix sources from auth and may introduce unintended side effects.

**Comment by facebook-github-bot-1** [comment]
@hmoqhim updated the pull request.

**Comment by drew-gross** [comment]
Looks good to me! Just making sure tests pass, then I will merge.

**Pull request merged**